python/tchannel/tornado/peer.py

```diff
@@ -458,35 +489,110 @@ def send(self, arg1, arg2, arg3, traceflag=False, headers=None):
+        blacklist = set()
+        # only retry on non-stream request
+        if is_stream(arg2, arg3):
+            connection = yield self.peer.connect()
+            message_id = connection.next_message_id()
+            original_request.id = message_id
+            response = yield connection.send_request(original_request)
+        else:
+            req = original_request
+            start_time = time.time()
+            try:
+                peer = self.peer
+                # mac number of times to retry 5.
+                for i in range(Request.MAX_RETRY_TIMES):
+                    connection = yield peer.connect()
+                    message_id = connection.next_message_id()
+                    req = original_request.clone()
+                    req.id = message_id
+                    try:
+                        time_left = Request.TIMEOUT - (
+                            time.time() - start_time)
+                        # request timeout 5s
+                        response_future = connection.send_request(req)
+                        with timeout(response_future, time_left):
+                            response = yield response_future
+                        break
+                    except ProtocolError as e:
+                        if not should_retry_on_error(req, e):
+                            raise
+                        blacklist.add(peer.hostport)
+
+                        # find new peer
+                        peer = peer.tchannel.peers.choose(
+                            hostport=self.r_hostport,
+                            score_threshold=self.r_score_threshold,
+                            blacklist=blacklist,
+                        )
+                        if not peer or i == Request.MAX_RETRY_TIMES - 1:
+                            raise
+            except TimeoutError:
+                req.set_exception(TimeoutError)
+                raise
+
         log.debug("Got response %s", response)
         raise gen.Return(response)
 
 
+def should_retry_on_error(request, error):
+    """rules for retry"""
+
+    retry_flag = request.headers['re']
+
+    if retry_flag == Retry.never:
+        return False
+
+    if error.code in [ErrorCode.bad_request, ErrorCode.cancelled]:
+        return False
+    elif error.code in [ErrorCode.busy, ErrorCode.declined]:
+        return True
+    elif error.code is ErrorCode.timeout:
+        return retry_flag is not Retry.connection_error
+    elif error.code in [ErrorCode.network_error, ErrorCode.unexpected]:
+        return retry_flag is not Retry.timeout
+    else:
+        log.error("unknown error type in request retry!")
+        return False
+
+
+def is_stream(arg2, arg3):
+    """check request is stream request or not"""
+    if (isinstance(arg2, InMemStream) and
+            isinstance(arg3, InMemStream) and
+            arg2.state == StreamState.completed and
+            arg3.state == StreamState.completed):
+        return False
+    else:
```

Comment:
Your'e not checking for PipeStream. 
`is_stream(PipeStream(..), PipeStream(..))` will be False.